include/vinecopulib/bicop/implementation/parametric.ipp

```diff
@@ -56,126 +56,126 @@ inline void ParBicop::fit(const Eigen::Matrix<double, Eigen::Dynamic, 2> &data,
                           double,
                           const Eigen::VectorXd& weights)
 {
-    if (family_ == BicopFamily::indep) {
-        // almost nothing to do
-        set_loglik(0.0);
-        return;
-    }
+      // for independence copula we don't have to do anything
+      if (family_ == BicopFamily::indep) {
+          set_loglik(0.0);
+          return;
+      }
 
-    using namespace tools_optimization;
-    std::vector <std::string> methods = {"itau", "mle"};
-    if (!tools_stl::is_member(method, methods)) {
-        throw std::runtime_error("Method not implemented.");
+      check_fit_method(method);
+      double tau = wdm::wdm(data, "tau", weights)(0, 1);
+
+      // for method itau and one-parameter families we don't need to optimize
+      int npars = static_cast<int>(calculate_npars()) - (method == "itau");
+      if (npars == 0) {
+          set_parameters(tau_to_parameters(tau));
+          set_loglik(loglik(data, weights));
+          return;
+      }
+
+      // Set bounds and starting values
+      auto lb = get_parameters_lower_bounds();
+      auto ub = get_parameters_upper_bounds();
+      adjust_parameters_bounds(lb, ub, tau, method);
+      auto initial_parameters = get_start_parameters(winsorize_tau(tau));
+
+      // find (pseudo-) mle
+      std::function<double(const Eigen::VectorXd&)> objective;
+      if (method == "mle") {
+          objective = [&data, &weights, this] (const Eigen::VectorXd& pars) {
+              this->set_parameters(pars);
+              return this->loglik(data, weights);
+          };
+      } else {
+          // profile likelihood
+          set_parameters(initial_parameters);
+          initial_parameters(0) = initial_parameters(1);
+          initial_parameters.conservativeResize(1);
+          objective = [&data, &weights, this] (const Eigen::VectorXd& pars) {
+              Eigen::VectorXd newpars(2);
+              newpars(0) = this->get_parameters()(0);
+              newpars(1) = pars(0);
+              this->set_parameters(newpars);
+              return this->loglik(data, weights);
+          };
+      }
+
+      tools_optimization::Optimizer optimizer;
+      auto newpars = optimizer.optimize(initial_parameters, lb, ub, objective);
+
+      // check if fit is reasonable, otherwise increase search interval
+      // and refit
+      if (optimizer.get_objective_max() < -0.1) {
```

Comment:
Aren't we doing the same optimization a second time for nothing for multiparameter families? If I understood the adjustment correctly, it's just for one-parameter families (for the Brent algorithm) no?

Reply:
Right, I'll fix that.